### PR TITLE
PERFORMANCE: Prefer concrete method over dynamic send for metric mutation.

### PR DESCRIPTION
**Pull request by jakelandis**
Addresses #7904

Lets wait for green before merge, I only ran a few of the key specs locally.

I will run some performance numbers here shortly.

### CODE REVIEW

**Comment by original-brownbear** [comment]
@jakelandis I'm seeing:

```
un'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:219:in `run'", "/Users/brownbear/src/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:132:in `run'", "/Users/brownbear/src/logstash/lib/bootstrap/environment.rb:71:in `<main>'"]}
[2017-08-04T16:29:13,913][ERROR][logstash.instrument.collector] Collector: Cannot create concrete class for this metric type {:type=>:gauge, :namespaces_path=>[:jvm, :memory, :pools, :survivor], :key=>:used_in_bytes, :metrics_params=>[:set, 35782656], :exception=>#<NoMethodError: undefined method `set' for nil:NilClass
Did you mean?  send>, :stacktrace=>["/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/metric_type/gauge.rb:16:in `set'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/collector.rb:48:in `push'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/metric.rb:32:in `gauge'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:144:in `block in collect_pools_metrics'", "org/jruby/RubyHash.java:1343:in `each'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:143:in `block in collect_pools_metrics'", "org/jruby/RubyHash.java:1343:in `each'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:142:in `collect_pools_metrics'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:56:in `collect'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/base.rb:57:in `start'", "org/jruby/RubyArray.java:2486:in `map'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_pollers.rb:23:in `start'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/agent.rb:419:in `configure_metrics_collectors'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/agent.rb:67:in `initialize'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:425:in `create_agent'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:338:in `execute'", "/Users/brownbear/src/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:67:in `run'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:219:in `run'", "/Users/brownbear/src/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:132:in `run'", "/Users/brownbear/src/logstash/lib/bootstrap/environment.rb:71:in `<main>'"]}
^C[2017-08-04T16:29:13,929][ERROR][logstash.instrument.collector] Collector: Cannot create concrete class for this metric type {:type=>:gauge, :namespaces_path=>[:jvm, :memory, :pools, :survivor], :key=>:committed_in_bytes, :metrics_params=>[:set, 35782656], :exception=>#<NoMethodError: undefined method `set' for nil:NilClass
Did you mean?  send>, :stacktrace=>["/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/metric_type/gauge.rb:16:in `set'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/collector.rb:48:in `push'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/metric.rb:32:in `gauge'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:144:in `block in collect_pools_metrics'", "org/jruby/RubyHash.java:1343:in `each'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:143:in `block in collect_pools_metrics'", "org/jruby/RubyHash.java:1343:in `each'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:142:in `collect_pools_metrics'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/jvm.rb:56:in `collect'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_poller/base.rb:57:in `start'", "org/jruby/RubyArray.java:2486:in `map'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/instrument/periodic_pollers.rb:23:in `start'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/agent.rb:419:in `configure_metrics_collectors'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/agent.rb:67:in `initialize'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:425:in `create_agent'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:338:in `execute'", "/Users/brownbear/src/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:67:in `run'", "/Users/brownbear/src/logstash/logstash-core/lib/logstash/runner.rb:219:in `run'", "/Users/brownbear/src/logstash/vendor/bundle/jruby/2.3.0/gems/clamp-0.6.5/lib/clamp/command.rb:132:in `run'", "/Users/brownbear/src/logstash/lib/bootstrap/environment.rb:71:in `<main>'"]}
```

File: logstash-core/lib/logstash/instrument/collector.rb [hunk 1]
```diff
@@ -37,7 +37,22 @@ def push(namespaces_path, key, type, *metric_type_params)
           LogStash::Instrument::MetricType.create(type, namespaces_path, key)
         end
 
-        metric.execute(*metric_type_params)
+        unless metric_type_params.size.eql? 2
+          logger.error("Collector: Cannot record metric. Unexpected metric params: " + metric_type_params.to_s)
+          return
+        end
+
+        method = metric_type_params[0]
```
Comment (original-brownbear) [inline logstash-core/lib/logstash/instrument/collector.rb]: @jakelandis this variable seems to only be used in the `else` branch, I think we can inline it there?

**Comment by jakelandis** [comment]
@original-brownbear - whoops this was bit pre-mature.  spec == passing != code works :( 

fixing...

File: logstash-core/lib/logstash/instrument/collector.rb (hunk 1 above)
Comment (jakelandis) [inline logstash-core/lib/logstash/instrument/collector.rb]: actually using it now.

**Comment by jakelandis** [comment]
@original-brownbear - OK this version should actually work.

**Comment by original-brownbear** [comment]
@jakelandis yea let's leave it out :( Tbh I like this version better still, but I'd suggest to not create any noise in this codebase, given that we want to quick-fix the performance here.
So if we don't backport this to all branches, it's just gonna make things weird (potentially).

**Comment by jakelandis** [comment]
@original-brownbear - agreed.

Closing.

**Pull request closed without merging**